velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,215 @@
+/// start and end are wrapped into the output ArrayVector.
+template <typename T>
+class ArrayDuplicatesFunction : public exec::VectorFunction {
+ public:
+  // Execute function.
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::Expr* caller,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indicesCursor;
+      vector_size_t numOfNulls = 0;
+
+      for (vector_size_t i = offset; i < offset + size; ++i) {
+        if (elements->isNullAt(i)) {
+          numOfNulls++;
+          if (numOfNulls == 2) {
+            rawNewIndices[indicesCursor] = i;
+            indicesCursor++;
+          }
+        } else {
+          T value = elements->valueAt<T>(i);
+          auto it = uniqueMap.find(value);
+          if (it == uniqueMap.end()) {
+            uniqueMap[value] = true;
+          } else if (it->second) {
+            it->second = false;
+            rawNewIndices[indicesCursor] = i;
+            indicesCursor++;
+          }
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indicesCursor - rawOffsets[row];
+
+      std::sort(
+          rawNewIndices + rawOffsets[row],
+          rawNewIndices + indicesCursor,
+          sparksql::Index2ValueNullableLess<T>(elements));
+    });
+
+    newIndices->setSize(indicesCursor * sizeof(vector_size_t));
+    auto newElements =
+        BaseVector::transpose(newIndices, std::move(elementsVector));
```

Comment:
If the input type is: `array(tinyint/smallint/int)`, the output type should be `array(bigint)`.

Do I need to explicitly create a new int64_t `ArrayVector` and copy all the values from `elementsVector`? If yes, is it possible to achieve that by borrowing the idea of `BaseVector.copy()` + `ArrayVector.copy()`?

Reply:
nvm, figured it out in c53da2c9d9fcdd29cec8a496db686cbbeb0d4010

Reply:
Per chatted with @mbasmanova , will limit Velox implementation to inputs of bigint and varchar only for now until we sort out the signature on the Presto side.